diff -u -r linux-source-4.3.orig/drivers/i2c/busses/i2c-designware-platdrv.c linux-source-4.3/drivers/i2c/busses/i2c-designware-platdrv.c
--- linux-source-4.3.orig/drivers/i2c/busses/i2c-designware-platdrv.c	2015-12-15 05:41:43.000000000 +0000
+++ linux-source-4.3/drivers/i2c/busses/i2c-designware-platdrv.c	2016-01-03 01:05:44.506366209 +0000
@@ -213,6 +213,8 @@
 			clk_freq = pdata->i2c_scl_freq;
 	}
 
+	ht = 230;
+
 	r = i2c_dw_eval_lock_support(dev);
 	if (r)
 		return r;
